Detectors/GlobalTracking/src/MatchTOF.cxx

```diff
@@ -1012,12 +1022,49 @@ void MatchTOF::doMatching(int sec)
       int eventIdTOF;
       int sourceIdTOF;
       for (auto iPropagation = 0; iPropagation < nStripsCrossedInPropagation; iPropagation++) {
+        int bct0 = int((mTOFClusWork[cacheTOF[itof]].getTime() - trkLTInt[iPropagation].getTOF(0) + 5000) * Geo::BC_TIME_INPS_INV); // bc taken assuming speed of light (el) and 5 ns of margin
+        if (bct0 < 0) {                                                                                                             // if negative time (it can happen at the beginng of the TF int was truncated per excess... adjusting)
+          bct0--;
+        }
+        float tof = mTOFClusWork[cacheTOF[itof]].getTime() - bct0 * Geo::BC_TIME_INPS;
+        if (tof - trkLTInt[iPropagation].getTOF(6) > 2000) { // reject tracks slower than triton
+          continue;
+        }
+        float cosangle = TMath::Cos(Geo::getAngles(indices[1], indices[2]) * TMath::DegToRad());
+        const float errXinvMin = 1. / (mMatchParams->maxResX * mMatchParams->maxResX);
+        const float errZinvMin = 1. / (mMatchParams->maxResZ * mMatchParams->maxResZ);
+        float errXinv2 = 1. / (trefTrk.getSigmaY2());
+        float errZinv2 = 1. / (trefTrk.getSigmaZ2() * cosangle); // should be valid only at eta=0
+
+        if (errXinv2 < errXinvMin) {
+          errXinv2 = errXinvMin;
+        }
+        if (errZinv2 < errZinvMin) {
+          errZinv2 = errZinvMin;
+        }
+
         LOG(debug) << "TOF Cluster [" << itof << ", " << cacheTOF[itof] << "]:      indices   = " << indices[0] << ", " << indices[1] << ", " << indices[2] << ", " << indices[3] << ", " << indices[4];
         LOG(debug) << "Propagated Track [" << itrk << "]: detId[" << iPropagation << "]  = " << detId[iPropagation][0] << ", " << detId[iPropagation][1] << ", " << detId[iPropagation][2] << ", " << detId[iPropagation][3] << ", " << detId[iPropagation][4];
         float resX = deltaPos[iPropagation][0] - (indices[4] - detId[iPropagation][4]) * Geo::XPAD + posCorr[0]; // readjusting the residuals due to the fact that the propagation fell in a pad that was not exactly the one of the cluster
         float resZ = deltaPos[iPropagation][2] - (indices[3] - detId[iPropagation][3]) * Geo::ZPAD + posCorr[2]; // readjusting the residuals due to the fact that the propagation fell in a pad that was not exactly the one of the cluster
         float res = TMath::Sqrt(resX * resX + resZ * resZ);
 
+        if (resX < -1.25) { // distance from closest border
+          resX += 1.25;
+        } else if (resX > 1.25) {
+          resX -= 1.25;
+        } else {
+          resX = 1E-3 / (pt + 1E-3); // high-pt should be favoured
+        }
+
+        if (resZ < -1.75) { // distance from closest border
+          resZ += 1.75;
+        } else if (resZ > 1.75) {
+          resZ -= 1.75;
+        } else {
+          resZ = 1E-3 / (pt + 1E-3); // high-pt should be favoured
+        }
```

Comment:
The logic here is that.
1) if a tracks enter in a pad  is always the best condition and it should be always accepted (residuals doesn't play a discriminating role here).
2) In case of two tracks entering in a pad I had to define a criterium to decide what is the right matching. Fastest track (higher pt) is favored since in a case of double hits only the first one is registered.
It is true that I loose the chi2 information but since the track entered in the pad it is not really useful.
What I am also interested to have is a Chi2 > 0 (even if by an epsilon) to distinguish such cases from the double hit case (at AO2D level).

Reply:
thanks, makes sense.